File: algorithms/sjf/sjf_sort.nim
import strformat, sugar, algorithm

type
  Process = object
    pid: Natural
    at: Natural
    bt: Natural
    ct: Natural

var prc = @[Process(pid: 0, at: 5, bt: 3, ct: 0), Process(pid: 1, at: 2, bt: 1, ct: 0),
            Process(pid: 2, at: 0, bt: 1, ct: 0), Process(pid: 3, at: 4, bt: 2, ct: 0),
            Process(pid: 4, at: 0, bt: 2, ct: 0), Process(pid: 5, at: 2, bt: 3, ct: 0)]
var ta, w = 0.0

proc max(x: openArray[Process]): int =
  result = x[0].bt
  for i in 1..high(x):
    if result < x[i].bt: result = x[i].bt

proc min(x: openArray[Process]): Process = 
  result = x[0]
  for i in 1..high(x):
    if result.at > x[i].at: result = x[i]

proc `==`[T](x, y: openArray[T]): bool =
  if x.pid == y.pid: true
  else: false

proc findCompletionTime(): void =
  prc[0].ct = prc[0].at + prc[0].bt
  var
    maxBT = max(prc)
    time = prc[0].ct
    count = 1
  while count < len(prc):
    var
      idx = -1
      minBT = maxBT
    for p in prc:
      if (p.ct == 0) and (p.bt <= minBT) and (p.at <= time):
        if idx != -1 and
          p.bt == minBT and p.at >= prc[idx].at:
            continue
        minBT = p.bt
        idx = find(prc, p)
    if idx == -1:
      var p = min(collect(newSeq, (for p in prc:
        if p.at > time: p)))
      time = p.at
      for i in 1..high(prc):
        if p.pid == prc[i].pid:
          idx = i
    time += prc[idx].bt
    prc[idx].ct = time
    count += 1

prc.sort(proc (x, y: Process): int =
  if x.bt > y.bt: 1
  else: -1
)
prc.sort(proc (x, y: Process): int =
  if x.at > y.at: 1
  else: -1
)
findCompletionTime()
prc.sort(proc (x, y: Process): int =
  if x.pid > y.pid: 1
  else: 0
)

echo "PID\tAT\tBT\tCT\tTAT\tWT"
for p in prc:
  let pTA = (float)(p.ct - p.at)
  let pW = pTA - (float)(p.bt)
  echo &"{p.pid}\t{p.at}\t{p.bt}\t{p.ct}\t{(int)pTA}\t{(int)pW}"
  ta += pTA / (float)(len(prc))
  w += pW / (float)(len(prc))

echo &"Avg. TAT: {ta:.3f}"
echo &"Avg. WT: {w:.3f}"

File: algorithms/sjf/sjf-sort.py
def findCompletionTime():
    proc[0]['ct'] = proc[0]['at'] + proc[0]['bt']
    time, maxBT = proc[0]['ct'], max(proc, key = lambda x: x['bt'])['bt']
    end = 1
    while end < len(proc):
        idx, minBT = -1, maxBT
        for i in range(1, len(proc)):
            if not proc[i]['ct'] and proc[i]['bt'] <= minBT and proc[i]['at'] <= time:
                if idx != -1 and proc[i]['bt'] == minBT and \
                        proc[i]['at'] >= proc[idx]['at']:
                            continue
                minBT = proc[i]['bt']
                idx = i
        if idx == -1:
            p = min([p for p in proc if p['at'] > time],
                    key = lambda x: x['at'])
            time, idx = p['at'], [i for i in range(1, len(proc)) if proc[i] is p][0]
        time += proc[idx]['bt']
        proc[idx]['ct'] = time
        end += 1

def findAvgTime():
    print('Avg TAT: ', end = '')
    print('%.3f'%(sum((x['ct'] - x['at']) for x in proc) / len(proc)))
    print('Avg WT: ', end = '')
    print('%.3f'%(sum((x['ct'] - (x['at'] + x['bt'])) for x in proc) / len(proc)))

if __name__ == '__main__':
    proc = [
            {'pid': 0, 'at': 13, 'bt': 2, 'ct': 0},
            {'pid': 1, 'at': 5, 'bt': 1, 'ct': 0},
            {'pid': 2, 'at': 2, 'bt': 3, 'ct': 0},
            {'pid': 3, 'at': 1, 'bt': 1, 'ct': 0},
            {'pid': 4, 'at': 0, 'bt': 4, 'ct': 0}
            ]

    proc.sort(key = lambda x: x['bt'])
    proc.sort(key = lambda x: x['at'])
    findCompletionTime()
    proc.sort(key = lambda x: x['pid'])

    print('PID\tAT\tBT\tCT\tTAT\tWT')
    [print(f"{p['pid']}",
        f"{p['at']}",
        f"{p['bt']}",
        f"{p['ct']}",
        f"{p['ct'] - p['at']}",
        f"{p['ct'] - (p['at'] + p['bt'])}",
        sep = '\t') for p in proc]
    findAvgTime()
